### PR TITLE
Implement rollup protocol

### PR DESCRIPTION
**Pull request by louismerlin**
**What this PR does**

*See commit message*
This PR implements the rollup protocol to replace the collect protocol, in the context of our master semester project.

---

🙅‍ Friendly checklist:

- [ ] 0. Code comments are added (or updated) when/where needed and explain the WHY of the code.
- [ ] 1. Design choices, user documentation and any additional doc are added (or updated) in READMEs.
- [ ] 2. Any new behaviour is tested and small units of code that can be are unit tested.
- [ ] 3. Code comments are added on tests to explain what they do.
- [ ] 4. Errors are systematically wrapped with a meaningful message using `xerrors.Errorf` and the `%v` verb.
- [ ] 5. Hard limit of 80 chars is always respected.
- [ ] 6. Changes are backward compatible.
- [ ] 7. Indentation level does not exceed 5, although 4 is already suspicious.
- [ ] 8. Functions, files, and packages are kept to a manageable size and decomposed into smaller units if needed.
- [ ] 9. There are no magic values.

### CODE REVIEW

File: byzcoin/service.go [hunk 1]
```diff
@@ -427,14 +423,76 @@ func (s *Service) AddTransaction(req *AddTxRequest) (*AddTxResponse, error) {
...
+
+		root := proto.(*RollupTxProtocol)
+		root.SkipchainID = req.SkipchainID
+		root.LatestID = latest.Hash
+		root.NewTx = req
+		err = proto.Start()
+		if err != nil {
+			return nil, xerrors.Errorf("Error starting the protocol", err)
+		}
+		if err := <-root.DoneChan; err != nil {
```
Comment (cgrigis) [inline byzcoin/service.go:474]: Perhaps it would be safer to `select` with a timeout, to avoid waiting forever

File: byzcoin/service.go [hunk 2]
```diff
@@ -445,10 +503,10 @@ func (s *Service) AddTransaction(req *AddTxRequest) (*AddTxResponse, error) {
 			return nil, xerrors.Errorf("couldn't get block info: %v", err)
 		}
 
-		ctxHash := req.Transaction.Instructions.Hash()
 		ch := s.notifications.registerForBlocks()
 		defer s.notifications.unregisterForBlocks(ch)
 
+		//TODO : create new block if txBuffer is not empty directly after creating another one
```
Comment (cgrigis) [inline byzcoin/service.go:509]: Please do :)

File: byzcoin/service.go [hunk 3]
```diff
@@ -2369,13 +2423,14 @@ func (s *Service) processOneTx(sst *stagingStateTrie, tx ClientTransaction,
...
 		copy(tx.Instructions[i+1+len(newInstructions):], tx.Instructions[i+1:])
 		copy(tx.Instructions[i+1:], newInstructions)
-
 		if err = sst.StoreAll(counterScs); err != nil {
 			err = xerrors.Errorf("%s StoreAll failed to add counter changes: %v",
 				s.ServerIdentity(), err)
 			s.addError(tx, err)
 			return nil, nil, err
 		}
+		//sstStoreAll.Record()
```
Comment (cgrigis) [inline byzcoin/service.go:2432]: to be removed?

File: byzcoin/viewchange_test.go [hunk 4]
```diff
@@ -20,6 +20,7 @@ import (
 // followers. Finally, we bring the failed nodes back up and they should
 // contain the transactions that they missed.
 func TestViewChange_Basic(t *testing.T) {
+	log.SetDebugVisible(2)
```
Comment (cgrigis) [inline byzcoin/viewchange_test.go:23]: to be removed after debug

File: byzcoin/tx_pipeline.go [hunk 5]
```diff
@@ -474,6 +412,8 @@ func (p *txPipeline) processTxs(initialState *txProcessorState) {
 					log.Lvl3("stopping txs processor")
 					return
 				}
+				//log.Print("received new tx from service", len(tx.Instructions), ok)
```
Comment (cgrigis) [inline byzcoin/tx_pipeline.go:415]: to be removed

File: byzcoin/rollup_tx_test.go [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+	"github.com/dedis/cothority_template/protocol"
+
+	"github.com/stretchr/testify/require"
+	"go.dedis.ch/onet/v3"
+	"go.dedis.ch/onet/v3/log"
+	"go.dedis.ch/onet/v3/network"
+)
+
+//var tSuite = suites.MustFind("Ed25519")
```
Comment (cgrigis) [inline byzcoin/rollup_tx_test.go:21]: to be removed

File: byzcoin/rollup_struct.go [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+const Name = "Rollup"
+
+// Announce is used to pass a message to all children.
+type Announce struct {
+	Message string
+}
+
+// announceWrapper just contains Announce and the data necessary to identify
+// and process the message in onet.
+type announceWrapper struct {
```
Comment (tharvik) [inline byzcoin/rollup_struct.go:22]: Unused (and Announce also)

File: byzcoin/rollup_struct.go [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+
+// Reply returns the count of all children.
+type Reply struct {
+	ChildrenCount int
+	Message       string
+}
+
+// replyWrapper just contains Reply and the data necessary to identify and
+// process the message in onet.
+type replyWrapper struct {
```
Comment (tharvik) [inline byzcoin/rollup_struct.go:35]: Also unused

File: byzcoin/rollup_tx.go [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+	*onet.TreeNodeInstance
+	TxsChan           chan []ClientTransaction
+	NewTx             *AddTxRequest
+	CtxChan           chan ClientTransaction
+	CommonVersionChan chan Version
+	SkipchainID       skipchain.SkipBlockID
+	LatestID          skipchain.SkipBlockID
+	MaxNumTxs         int
+	DoneChan          chan error
+	getTxs            getTxsCallback
```
Comment (tharvik) [inline byzcoin/rollup_tx.go:37]: unused

File: byzcoin/rollup_tx.go [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+	"golang.org/x/xerrors"
+)
+
+func init() {
+	network.RegisterMessages(&AddTxRequest{}, &RequestAdded{})
+	_, err := onet.GlobalProtocolRegister(rollupTxProtocol, NewRollupTxProtocol)
+	log.ErrFatal(err)
+}
+
+type getTxsCallback func(*network.ServerIdentity, *onet.Roster,
```
Comment (tharvik) [inline byzcoin/rollup_tx.go:20]: unused

File: byzcoin/service_test.go [hunk 11]
```diff
@@ -45,7 +45,7 @@ const stateChangeCacheContract = "stateChangeCacheTest"
 
 func TestMain(m *testing.M) {
 	log.SetShowTime(true)
-	log.MainTest(m)
+	log.MainTest(m, 1)
```
Comment (tharvik) [inline byzcoin/service_test.go:48]: debug

File: byzcoin/rollup_tx.go [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+	return p.SendToParent(&RequestAdded{})
+}
+
+// Shutdown closes the closing channel to abort any waiting on messages.
+func (p *RollupTxProtocol) Shutdown() error {
+	close(p.closing)
+	return nil
+}
+
+func (p *RollupTxProtocol) getByzcoinVersion() Version {
```
Comment (tharvik) [inline byzcoin/rollup_tx.go:129]: unused

File: byzcoin/service.go (hunk 1 above)
Comment (louismerlin) [reply byzcoin/service.go]: This was changed

File: byzcoin/service.go [hunk 13]
```diff
@@ -427,14 +423,81 @@ func (s *Service) AddTransaction(req *AddTxRequest) (*AddTxResponse, error) {
 		log.Lvlf2("Instruction[%d]: %s on instance ID %s", i, instr.Action(), instr.InstanceID.String())
 	}
 
+	//Either send the transaction to the leader, or, if this node is the leader, directly send it to ctxChan.
+	//For every new tx create a new protocol, like in skipchain
+
+	leader, err := s.getLeader(req.SkipchainID)
+	if err != nil {
+		return nil, xerrors.Errorf("Error getting the leader", err)
```
Comment (tharvik) [inline byzcoin/service.go:431]:
```suggestion
		return nil, xerrors.Errorf("Error getting the leader: %w", err)
```

File: byzcoin/service.go [hunk 14]
```diff
@@ -427,14 +423,81 @@ func (s *Service) AddTransaction(req *AddTxRequest) (*AddTxResponse, error) {
...
+			return nil, xerrors.Errorf("creating protocol: %v", err)
+		}
+
+		root := proto.(*RollupTxProtocol)
+		root.SkipchainID = req.SkipchainID
+		root.LatestID = latest.Hash
+		root.NewTx = req
+		err = proto.Start()
+		if err != nil {
+			return nil, xerrors.Errorf("Error starting the protocol", err)
```
Comment (tharvik) [inline byzcoin/service.go:471]:
```suggestion
			return nil, xerrors.Errorf("Error starting the protocol: %w", err)
```

**Comment by ineiti** [comment]
Closed in favour of #2321

**Comment by louismerlin** [comment]
Sorry about that ! We've lacked the motivation to finish the PR now that we're both at 100% employment... We hope the new one will get through :)

**Pull request closed without merging**